torch/_inductor/optimize_indexing.py

```diff
@@ -136,77 +156,105 @@ def constant(value, dtype):
 
     @staticmethod
     def reciprocal(x):
-        return ValueRanges.checked_unary_map(x, lambda y: 1 / y)
+        x = ValueRanges.wrap(x)
+        if 0 in x:
+            return ValueRanges(-math.inf, math.inf)
+        else:
+            return ValueRanges.decreasing_map(x, lambda y: 1 / y)
+
+    @staticmethod
+    def square(x):
+        return ValueRanges.convex_min_zero_map(x, lambda y: y * y)
 
     @staticmethod
     def abs(x):
-        return ValueRanges.checked_unary_map(x, abs)
+        return ValueRanges.convex_min_zero_map(x, abs)
 
     @staticmethod
     def neg(x):
-        return ValueRanges.checked_unary_map(x, lambda x: -x)
+        return ValueRanges.decreasing_map(x, operator.neg)
 
     @staticmethod
     def truediv(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.truediv)
+        b = ValueRanges.wrap(b)
+        if 0 in b:
+            return ValueRanges(-math.inf, math.inf)
+        else:
+            return ValueRangeAnalysis.mul(a, ValueRanges(1 / b.upper, 1 / b.lower))
 
     @staticmethod
     def div(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.truediv)
+        # We think of this as floor(a / b)
+        out = ValueRangeAnalysis.truediv(a, b)
+        floor = sympy.functions.elementary.integers.floor
+        return ValueRanges(floor(out.lower), floor(out.upper))
```

Comment:
I dont think this is what we want if the inputs are floating point. could you also check that sympy derivatives are able to handle floor ?

Reply:
I restructured a bit this function and `floor` and `ceil` for their implementation to be a bit simpler and correct.